gapis/api/vulkan/transform_command_splitter.go

```diff
@@ -24,961 +24,1067 @@ import (
+	currentRenderPassArgs VkCmdBeginRenderPassXArgsʳ,
+	newRenderPass VkRenderPass) (VkCmdBindPipelineArgsʳ, error) {
+
+	// We are only interested in graphics pipeline
+	if bindPipelineArgs.PipelineBindPoint() != VkPipelineBindPoint_VK_PIPELINE_BIND_POINT_GRAPHICS {
+		return bindPipelineArgs, nil
+	}
+
+	stateObject := GetState(inputState)
 
-	commandPoolID, err := splitTransform.getCommandPool(ctx, queueSubmit, inputState)
+	// If we have only one renderpass we don't modify the renderpass, therefore no change is required.
+	if stateObject.RenderPasses().Get(currentRenderPassArgs.RenderPassBeginInfo().RenderPass()).SubpassDescriptions().Len() <= 1 {
+		return bindPipelineArgs, nil
+	}
+
+	newPipeline, err := splitter.rewriteGraphicsPipeline(ctx, inputState,
+		bindPipelineArgs.Pipeline(), newRenderPass)
 	if err != nil {
-		log.E(ctx, "Failed during getting command pool : %v", err)
+		log.E(ctx, "Failed during rewriting graphics pipeline : %v", err)
+		return VkCmdBindPipelineArgsʳ{}, err
+	}
+
+	newPipelineArgs := bindPipelineArgs.Clone(api.CloneContext{})
+	newPipelineArgs.SetPipeline(newPipeline)
+	return newPipelineArgs, nil
+}
+
+type renderPassRequirements struct {
+	loadAttachment                 bool
+	storeAttachment                bool
+	updateFinalLayout              bool
+	updateSubpassLayoutTransitions bool
+	preventAttachmentResolve       bool
+}
+
+func (splitter *commandSplitter) createRenderPassSplitters(ctx context.Context, inputState *api.GlobalState, renderPassObject RenderPassObjectʳ) renderPassSplitters {
+	st := GetState(inputState)
+
+	// Check if we already split this renderpass
+	if splitRenderPasses, ok := splitter.splitRenderPasses[renderPassObject.VulkanHandle()]; ok {
+		return splitRenderPasses
+	}
+
+	handles := make(renderPassSplitters, 0)
+	currentLayouts := make(map[uint32]VkImageLayout)
+
+	for i := uint32(0); i < uint32(renderPassObject.AttachmentDescriptions().Len()); i++ {
+		currentLayouts[i] = renderPassObject.AttachmentDescriptions().Get(i).InitialLayout()
+	}
+
+	tempTransformWriter := newCommandSplitTransformWriter(inputState, splitter)
+	sb := st.newStateBuilder(ctx, newTransformerOutput(tempTransformWriter))
+
+	// For every subpass we have, we create 3 kind of renderpass. How they will be used at the end
+	// is based on what commands requested.
+	//
+	// For example for a renderpass that has 3 subpass with 3 draw in each:
+	// {
+	//	SP1[draw, draw, draw], SP2[draw, draw, draw], SP3[draw, draw, draw]
+	// }
+	//
+	// Let's assume we request drawcalls with *
+	// {
+	// SP1[draw*, draw, draw], SP2[draw, draw*, draw], SP3[draw, draw*, draw*]
+	// }
+	//
+	// The new renderpasses at the end of this transform will be
+	// {
+	// Enter1[draw*], [InsertionCmd], Intermediate1[draw, draw], Exit1[],
+	// Enter2[draw, draw*], [InsertionCmd], Intermedite2[draw], Exit2[],
+	// Enter3[draw, draw*], [InsertionCmd], Intermediate2[draw*], [InsertionCmd], Intermediate2[], Exit2[]
+	// }
```

Comment:
This example is really nice to help understand what the command splitter is doing, but it fails to give an example where ExitN[] is non empty. That would really help understand why the exit subpass exists!

Reply:
Oh, I see the answer to my own comment below. It's basically always empty and just used for tile format transitions. Do we always need to have it if it's not doing anything? I guess it makes the code simpler to always insert one even if it is a noop. 

This function is *really* nice. It makes a high level picture of what is happening with this transform much more readable.

Reply:
Thanks!